gsdk/src/signer/mod.rs

```diff
@@ -42,32 +75,72 @@ pub struct Signer {
 impl Signer {
     /// New signer api.
     pub fn new(api: Api, suri: &str, passwd: Option<&str>) -> Result<Self> {
-        Ok(Self {
+        let signer = SignerInner {
             api,
             signer: PairSigner::new(
                 Pair::from_string(suri, passwd).map_err(|_| Error::InvalidSecret)?,
             ),
             nonce: None,
-        })
+        };
+
+        Ok(Self::from_inner(signer))
+    }
+
+    fn from_inner(signer: SignerInner) -> Self {
+        let signer = Arc::new(signer);
+
+        Self {
+            sudo: SignerSudo(signer.clone()),
+            balance: SignerBalance(signer.clone()),
+            calls: SignerCalls(signer.clone()),
+            rpc: SignerRpc(signer.clone()),
+            signer,
+        }
+    }
```

Comment:
since we are using `Arc` now, don't have to clone everything again, better find a way applying the changes via just updating the inner